### ISSUE TITLE
Incorrect throttling behavior

### ISSUE THREAD

**Issue by eimantas** [pointfreeco/swift-composable-architecture#540]
I know `Effect.throttle` is not yet public, and as per [this thread](https://forums.swift.org/t/type-inference-when-using-effect-throttle-operator/47994) I decided to "kick the tires". My first try was with some interesting behavior involving composable core location. I will try to make a complete example, but cannot guarantee it.

**To Reproduce**
Create a location manager through `ComposableCoreLocation` and `set` custom properties. Handle `didUpdateLocations` event from `LocationManager.Action` in your own reducer and return effect with a different event that, say, should push the coordinates to backend api every 15 seconds (the throttling part).

```swift
struct AppState {
    var latestLocations: [Location] = []
}

enum AppAction {
    case startUpdating
    case stopUpdating
    case locationManager(LocationManager.Action)
    case reportToApi
}

struct AppEnvironment {
    var managerId = "ManagerId"
    var locationManager: LocationManager = .live
    var scheduler: AnySchedulerOf<DispatchQueue>
    var reportId = "ReportId"
}

let appReducer = Reducer<AppState,AppAction,AppEnvironment> { state,action,env in

    switch action {
    case .stopUpdating:
        return .merge(
            env.locationManager.stopUpdatingLocation(id: env.managerId)
                .fireAndForget(),
            env.locationManager.destroy(id: env.managerId)
                .fireAndForget()
        )
    case .startUpdating:
        return .merge(
            env.locationManager.create(id: env.managerId)
                .map(AppAction.locationManager),
            env.locationManager.set(
                    id: env.managerId,
                    activityType: .other,
                    allowsBackgroundLocationUpdates: true,
                    desiredAccuracy: kCLLocationAccuracyNearestTenMeters
                    distanceFilter: 40
                )
                .fireAndForget(),
            env.locationManager.startUpdatingLocation(id: env.managerId)
                .fireAndForget()
        )
    case .locationManager(.didUpdateLocations(let locations)):
        state.latestLocations = locations
        // this is where the throttling happens
        return Effect(value: .reportLocation)
            .throttle(
                id: env.reportId,
                for: 15,
                scheduler: env.scheduler,
                latest: true
            )
    case .reportLocation:
        return .fireAndForget {
            print("[.reportLocation] \(Date().debugDescription)")
        }
    }
}
```

**Expected behavior**
I expected to see printed dates every 15 seconds.

**Screenshots**
Not screenshots, but actual console output:

```log
[.reportLocation] 2021-05-10 12:53:44 +0000
[.reportLocation] 2021-05-10 12:53:59 +0000
[.reportLocation] 2021-05-10 12:54:00 +0000
[.reportLocation] 2021-05-10 12:54:15 +0000
[.reportLocation] 2021-05-10 12:54:30 +0000
[.reportLocation] 2021-05-10 12:54:45 +0000
[.reportLocation] 2021-05-10 12:54:45 +0000
[.reportLocation] 2021-05-10 12:55:00 +0000
[.reportLocation] 2021-05-10 12:55:01 +0000
[.reportLocation] 2021-05-10 12:55:16 +0000
[.reportLocation] 2021-05-10 12:55:16 +0000
[.reportLocation] 2021-05-10 12:55:31 +0000
[.reportLocation] 2021-05-10 12:55:32 +0000
[.reportLocation] 2021-05-10 12:55:47 +0000
[.reportLocation] 2021-05-10 12:55:47 +0000
```

**Environment**
 - Xcode 12.5
 - watchOS 7.4

**Additional context**
Business logic is to track user's location and report it every 15 seconds to the API. The `startUpdating` and `stopUpdating` can be hooked to simple button in UI. To achieve the above output I used "City Run" simulated location on watchOS simulator.

**Comment by stephencelis** [comment]
@eimantas I meant to reply earlier but had been busy at the time. Would you happen to still have this project available to attach (view, etc.)?

**Comment by kerryknight** [comment]
Hi,

*TLDR: Shouldn't the delayed value publisher (`#4` in the implementation of `throttle` below from `Throttling.swift`) also be updating the cached throttle time (`#5`) once it publishes its value, so that the throttle time interval window is reset?* 

-- 

I *believe* the issue might be that the cached throttle time (`throttleTime`) is not updated whenever the delayed value is published. Without updating the cached throttle time there as well, the interval window is not reset so if values continue to come in (like a new location update is received or a user is button spamming), `#2` will still fire. This results in multiple values getting sent within the same interval window unexpectedly.  

However, if the delayed value publisher (`#4`) updates the cached throttle time once the delayed value is sent (`#5`), the `#2` value publisher will be bypassed as expected if a new value comes in before the end of the next interval duration.  Note, I believe this code also sends the *last* value that was received within the specified time interval, which I think is similar to the behavior of `throttle` in RxSwift (but not necessarily Combine?).

*(Please note, I tested all this behavior in my RxSwift implementation using the RxSwift TCA fork. I've ported my changes for that to Combine here for demo purposes but have not tested them directly.)*

```
  // snippet pasted from Throttling.swift with comments added by me
  func throttle<S>(
    id: AnyHashable,
    for interval: S.SchedulerTimeType.Stride,
    scheduler: S,
    latest: Bool
  ) -> Effect where S: Scheduler {
    self.flatMap { value -> AnyPublisher<Output, Failure> in
      guard let throttleTime = throttleTimes[id] as! S.SchedulerTimeType? else {
        // #1: update cached throttle time to .now and publish value
        throttleTimes[id] = scheduler.now
        throttleValues[id] = nil
        return Just(value).setFailureType(to: Failure.self).eraseToAnyPublisher()
      }

      guard throttleTime.distance(to: scheduler.now) < interval else {
        // #2: cached throttle time elapsed, update it to .now and publish value
        throttleTimes[id] = scheduler.now
        throttleValues[id] = nil
        return Just(value).setFailureType(to: Failure.self).eraseToAnyPublisher()
      }

      // #3: shouldn't these two lines go before the 2nd guard statement above? 
      // Otherwise, #2 above will always simply return the latest value, correct?
      let value = latest ? value : (throttleValues[id] as! Output? ?? value)
      throttleValues[id] = value

      // #4: delay publishing the value (note, cached throttle time
      // is not currently updated when the value is eventually sent)
      return Just(value)
        .delay(
          for: scheduler.now.distance(to: throttleTime.advanced(by: interval)), scheduler: scheduler
        )
        // < ------------ MY ADDITION ------------- >
        .handleEvents(receiveOutput: { [weak self] _ in
            // #5: update the cached throttle time to .now here as well, which 
            // has the effect of resetting the interval window so that #2 won't 
            // fire if new values come in quickly afterwards
            throttleTimes[id] = scheduler.now
        })
        // </ -------_--- MY ADDITION ------------- >
        .setFailureType(to: Failure.self)
        .eraseToAnyPublisher()
    }
    .eraseToEffect()
    .cancellable(id: id, cancelInFlight: true)
  }
```

**Comment by stephencelis** [comment]
@kerryknight Thanks for the very detailed explanation! I've submitted a PR here with your fix and a test in which I was able to reproduce the bad behavior: https://github.com/pointfreeco/swift-composable-architecture/pull/654

>       // #3: shouldn't these two lines go before the 2nd guard statement above? 
>       // Otherwise, #2 above will always simply return the latest value, correct?
>       let value = latest ? value : (throttleValues[id] as! Output? ?? value)
>       throttleValues[id] = value

Good catch! That absolutely was a bug, which I was able to reproduce in a failing test, thanks!

If you'd like to submit the PR with the original fix, we'll happily merge it instead, and I can either merge my test separately, or you can include it in your PR. If you don't have the time or interest in submitting your own PR, just let us know, and we'll merge the one I opened instead 😄

**Comment by kerryknight** [comment]
Thanks @stephencelis! I won't be able to get back to this to open a PR for a couple days, so I have no problem if you just want to go ahead with your own PR. Otherwise, I can get the PR opened by the end of this week. Up to you. Thanks for the great work!

**Comment by eimantas** [comment]
Sorry for not responding to this thread sooner as my github inbox is bombarded with work-related notifications >.<

Thank you @kerryknight for a thorough investigation and thank you @stephencelis for a fix!

❤️

**Comment by stephencelis** [comment]
@eimantas Please let us know if you can verify things work as they should now! 😄 Would hate to roll out 0.22.0 and find that it's still broken 😆
